jni/VuforiaNative.cpp

```diff
@@ -0,0 +1,607 @@
+    
+    // Get the image tracker:
+    QCAR::TrackerManager& trackerManager = QCAR::TrackerManager::getInstance();
+    QCAR::ObjectTracker* imageTracker = static_cast<QCAR::ObjectTracker*>(
+                    trackerManager.getTracker(QCAR::ObjectTracker::getClassType()));
+    if (imageTracker == NULL)
+    {
+        LOG("Failed to load tracking data set because the ImageTracker has not"
+            " been initialized.");
+        return 0;
+    }
+
+    // Create the data sets:
+    dataSetStonesAndChips = imageTracker->createDataSet();
+    if (dataSetStonesAndChips == 0)
+    {
+        LOG("Failed to create a new tracking data.");
+        return 0;
+    }
+
+    // Load the data sets:
+    if (!dataSetStonesAndChips->load("AR4AndroidApplication.xml", QCAR::DataSet::STORAGE_APPRESOURCE))
+    {
+        LOG("Failed to load data set.");
+        return 0;
+    }
+
+    // Activate the data set:
+    if (!imageTracker->activateDataSet(dataSetStonesAndChips))
+    {
+        LOG("Failed to activate data set.");
+        return 0;
+    }
+
+    LOG("Successfully loaded and activated data set.");
+    return 1;
+}
+
+
+JNIEXPORT int JNICALL
+Java_com_ar4android_AR4AndroidActivity_destroyTrackerData(JNIEnv *, jobject)
+{
+    LOG("Java_com_ar4android_AR4AndroidActivity_destroyTrackerData");
+
+    // Get the image tracker:
+    QCAR::TrackerManager& trackerManager = QCAR::TrackerManager::getInstance();
+    QCAR::ObjectTracker* imageTracker = static_cast<QCAR::ObjectTracker*>(
+        trackerManager.getTracker(QCAR::ObjectTracker::getClassType()));
+    if (imageTracker == NULL)
+    {
+        LOG("Failed to destroy the tracking data set because the ImageTracker has not"
+            " been initialized.");
+        return 0;
+    }
+    
+    if (dataSetStonesAndChips != 0)
+    {
+        if (imageTracker->getActiveDataSet() == dataSetStonesAndChips &&
+            !imageTracker->deactivateDataSet(dataSetStonesAndChips))
+        {
+            LOG("Failed to destroy the tracking data set StonesAndChips because the data set "
+                "could not be deactivated.");
+            return 0;
+        }
+
+        if (!imageTracker->destroyDataSet(dataSetStonesAndChips))
+        {
+            LOG("Failed to destroy the tracking data set StonesAndChips.");
+            return 0;
+        }
+
+        LOG("Successfully destroyed the data set StonesAndChips.");
+        dataSetStonesAndChips = 0;
+    }
+
+    return 1;
+}
+
+
+JNIEXPORT void JNICALL
+Java_com_ar4android_AR4AndroidActivity_onQCARInitializedNative(JNIEnv *, jobject)
+{
+    // Register the update callback where we handle the data set swap:
+    QCAR::registerCallback(&updateCallback);
+
+    // Comment in to enable tracking of up to 2 targets simultaneously and
+    // split the work over multiple frames:
+    // QCAR::setHint(QCAR::HINT_MAX_SIMULTANEOUS_IMAGE_TARGETS, 2);
+}
+
+// RENDERING CALL
+
+JNIEXPORT void JNICALL
```

Comment:
This function is used to query the position of the targets and transfer it to the Java layer